src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,311 @@
+ * All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "ofi_hmem.h"
+#include "ofi.h"
+
+#ifdef HAVE_GDRCOPY
+
+#include <gdrapi.h>
+
+struct gdrcopy_ops {
+
+	gdr_t (*gdr_open)();
+
+	int (*gdr_close)(gdr_t g);
+
+	int (*gdr_pin_buffer)(gdr_t g, unsigned long addr, size_t size,
+			      uint64_t p2p_token, uint32_t va_space,
+			      gdr_mh_t *handle);
+
+	int (*gdr_unpin_buffer)(gdr_t g, gdr_mh_t handle);
+
+	int (*gdr_map)(gdr_t g, gdr_mh_t handle, void **va, size_t size);
+
+	int (*gdr_unmap)(gdr_t g, gdr_mh_t handle, void *va, size_t size);
+
+	int (*gdr_copy_to_mapping)(gdr_mh_t handle, void *map_d_ptr,
+				   const void *h_ptr, size_t size);
+
+	int (*gdr_copy_from_mapping)(gdr_mh_t handle, void *map_d_ptr,
+				     const void *h_ptr, size_t size);
+};
+
+#ifdef ENABLE_GDRCOPY_DLOPEN
+
+#include <dlfcn.h>
+
+static void *gdrapi_handle;
+static struct gdrcopy_ops gdrcopy_ops;
+
+#else
+
+static struct gdrcopy_ops gdrcopy_ops = {
+	.gdr_open = gdr_open,
+	.gdr_close = gdr_close,
+	.gdr_pin_buffer = gdr_pin_buffer,
+	.gdr_unpin_buffer = gdr_unpin_buffer,
+	.gdr_map = gdr_map,
+	.gdr_unmap = gdr_unmap,
+	.gdr_copy_to_mapping = gdr_copy_to_mapping,
+	.gdr_copy_from_mapping = gdr_copy_from_mapping
+};
+
+#endif /* ENABLE_CUDA_DLOPEN */
+
+gdr_t ofi_gdrcopy_open()
+{
+	if (!gdrcopy_ops.gdr_open)
+		return NULL;
+
+	return gdrcopy_ops.gdr_open();
+}
+
+int ofi_gdrcopy_close(gdr_t gdr)
+{
+	assert(gdrcopy_ops.gdr_close);
+
+	return gdrcopy_ops.gdr_close(gdr);
+}
+
+ssize_t ofi_gdrcopy_reg(void *addr, size_t len, gdr_t gdr, struct ofi_gdrcopy_handle *gdrcopy)
```

Comment:
I wonder if we need a new ofi HMEM operation to support this use case. For example, I have been looking at implementing an ofi_hmem_host_register(). Maybe we should have a ofi_hmem_dev_register() which returns an opaque handle to the caller.

Reply:
Sure, I will do that.